pdf/pdfrenderthread.cpp

```diff
@@ -126,7 +127,28 @@ class PDFRenderThreadPrivate
             delete page;
         }
     }
-
+    void retrieveTextBoxes(int i)
+    {
+        if (i < 0 || i >= document->numPages()) {
+            return;
+        }
+        Poppler::Page* page = document->page(i);
+        QList<Poppler::TextBox*> plst = page->textList();
+        QList< QPair< QRectF, QString > > lst;
+        for (QList<Poppler::TextBox*>::const_iterator box = plst.begin();
+             box != plst.end(); box++) {
+            QSizeF psize = page->pageSize();
+            QRectF bbox = (*box)->boundingBox();
+            QRectF rect{ bbox.x() / psize.width(),
+                    bbox.y() / psize.height(),
+                    bbox.width() / psize.width(),
+                    bbox.height() / psize.height()};
+            lst.append(QPair<QRectF, QString>{rect, (*box)->text() + ((*box)->hasSpaceAfter() ? " " : "")});
```

Comment:
As you said, this will almost always append a space on copied text, which feels unfortunate. How about class for text box instead of QPair? Not fully sure what to think of using Poppler one. In practice it looks safe, though Poppler doesn't seem to specify thread safety nearly at all.

Btw, noticed that on Poppler operations, toc model calls some outside render thread.